pkg/volume/csi/csi_node_updater.go

```diff
@@ -0,0 +1,178 @@
+	startMu sync.Mutex
+}
+
+// NewCSINodeUpdater creates a new csiNodeUpdater
+func NewCSINodeUpdater(
+	driverLister storagelisters.CSIDriverLister, driverInformer cache.SharedIndexInformer) (*csiNodeUpdater, error) {
+	if driverLister == nil {
+		return nil, fmt.Errorf("driverLister must not be nil")
+	}
+	if driverInformer == nil {
+		return nil, fmt.Errorf("driverInformer must not be nil")
+	}
+	return &csiNodeUpdater{
+		driverLister:   driverLister,
+		driverInformer: driverInformer,
+		driverUpdaters: sync.Map{},
+	}, nil
+}
+
+// Run starts the CSINodeUpdater
+func (u *csiNodeUpdater) Run() {
+	u.startMu.Lock()
+	defer u.startMu.Unlock()
+
+	if u.started {
+		return
+	}
+
+	_, err := u.driverInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    u.onDriverAdd,
+		UpdateFunc: u.onDriverUpdate,
+		DeleteFunc: u.onDriverDelete,
+	})
+	if err != nil {
+		klog.ErrorS(err, "Failed to add event handler for CSI driver informer")
+		return
+	}
+
+	klog.InfoS("CSINodeUpdater initialized successfully")
+	u.started = true
+}
+
+// onDriverAdd handles the addition of a new CSIDriver object
+func (u *csiNodeUpdater) onDriverAdd(obj interface{}) {
+	driver, ok := obj.(*v1.CSIDriver)
+	if !ok {
+		return
+	}
+	u.startOrReconfigureDriverUpdater(driver)
+}
+
+// onDriverUpdate handles updates to CSIDriver objects
+func (u *csiNodeUpdater) onDriverUpdate(oldObj, newObj interface{}) {
+	oldDriver, ok := oldObj.(*v1.CSIDriver)
+	if !ok {
+		return
+	}
+	newDriver, ok := newObj.(*v1.CSIDriver)
+	if !ok {
+		return
+	}
+
+	// Check if the relevant field changed
+	oldPeriod := getNodeAllocatableUpdatePeriod(oldDriver)
+	newPeriod := getNodeAllocatableUpdatePeriod(newDriver)
+
+	if oldPeriod != newPeriod {
+		klog.InfoS("NodeAllocatableUpdatePeriodSeconds changed", "driver", newDriver.Name, "oldPeriod", oldPeriod, "newPeriod", newPeriod)
+		u.startOrReconfigureDriverUpdater(newDriver)
+	}
+}
+
+// onDriverDelete handles deletion of CSIDriver objects
+func (u *csiNodeUpdater) onDriverDelete(obj interface{}) {
+	driver, ok := obj.(*v1.CSIDriver)
+	if !ok {
+		return
+	}
+	u.unregisterDriver(driver.Name)
+}
+
+// unregisterDriver stops updates for a driver
+func (u *csiNodeUpdater) unregisterDriver(driverName string) {
+	klog.V(4).InfoS("UnregisterDriver called", "driver", driverName)
+
+	if stopCh, exists := u.driverUpdaters.Load(driverName); exists {
+		close(stopCh.(chan struct{}))
+		u.driverUpdaters.Delete(driverName)
+		klog.V(4).InfoS("Stopped updater for driver", "driver", driverName)
+	}
+}
+
+// startOrReconfigureDriverUpdater starts or reconfigures the updater for a driver
+func (u *csiNodeUpdater) startOrReconfigureDriverUpdater(driver *v1.CSIDriver) {
+	period := getNodeAllocatableUpdatePeriod(driver)
+
+	// If the period is 0, disable updates for this driver
+	if period == 0 {
+		klog.InfoS("NodeAllocatableUpdatePeriodSeconds is 0, disabling updates", "driver", driver.Name)
+		u.unregisterDriver(driver.Name)
+		return
+	}
+
```

Comment:
I think we should check here for installed driver via `csiDrivers` map, if CSI driver in question is really installed before starting the goroutine which will periodically poll the driver etc. This will save some CPU cycles. For example, may be cluster has windows nodes, where no driver gets installed but yet it receives `CSIDriver` object updates etc.

Reply:
Hmm, that would indeed save a few CPU cycles, but creates a new problem where the periodic update is never started because we lose the following race:

CSI Node updater sees driver hasn't been registered:
```
Mar 11 16:41:22 i-0fc49c77ef5fe4355.us-west-2.compute.internal kubelet[235658]: I0311 16:41:22.571627  235658 csi_node_updater.go:144] "CSI driver not installed on this node, skipping periodic updater" driver="ebs.csi.aws.com"
```

Driver is added to global map a few ms after CSI Node updater has already determined it should skip:
```
Mar 11 16:41:22 i-0fc49c77ef5fe4355.us-west-2.compute.internal kubelet[235658]: I0311 16:41:22.657749  235658 csi_plugin.go:130] "Added CSI Driver"
```

Reply:
Solved in latest revision!